File: Sept04/solution.cpp
class Solution {
public:
    static int firstStableIndex(vector<int>& nums, int k) {
        int min_val = INT_MAX;
        int max_val = INT_MIN;
        vector<int> suffixMin(nums.size(), 0);
        int inst = 0;

        for(int i=nums.size()-1; i>=0; i--)
        {
            min_val = min(min_val, nums[i]);
            suffixMin[i] = min_val;
        }
        
        for(int i=0; i<nums.size(); i++)
        {
            max_val = max(max_val, nums[i]);
            inst = max_val - suffixMin[i];

            if (inst <= k)
            {
                return i;
            }
        }

        return -1;
    }
};

File: Sept04/solution.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        min_val = 2**31-1
        max_val = -2**31

        suffixMin = [0]*len(nums)
        i = len(nums)-1
        while i >= 0:
            min_val = min(nums[i], min_val)
            suffixMin[i] = min_val
            i -=1
        
        i= 0
        for i in range(0, len(nums)):
            max_val = max(max_val, nums[i])
            instab = max_val - suffixMin[i]
            if instab <= k:
                return i
        
        return -1
